Implement File Locking

weirddan455 opened the pull request:
# Description

I've implemented locking the entire file with sharing bits in `DOS_OpenFile` and `DOS_CreateFile`.

There is another (completely independent from sharing bits) method that lock regions of a file.  This is done through INT 21 AH = 0x5c.  This requires performing a check on `DOS_ReadFile` and `DOS_WriteFile` to check if the region a process is attempting to read/write to has been locked by another process.

Many of the games I've tested require Win32s (who's installer does a `SHARE.EXE` check).  This check failed for me on both DOSBox-X and using `FAKESHR.COM` so we are doing better than them now :smile: 


## Related issues

Fixes #1489

# Manual testing

- 3-D Ultra Pinball
- Comix Zone
- Microsoft Flight Simulator 5 - ATC Workshop
- AFL Finals Fever
- Mind Grid - Install errors but probably not related to this PR (also producible on main see #3703)
- Microsoft Word 6.1 for Windows


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -603,6 +603,129 @@ bool DOS_CreateFile(const char* name, FatAttributeFlags attributes,
+	}
+
+	// Extract bits 4-6
+	const uint8_t sharing_bits = (flags & 0b1110000) >> 4;
```
This is a perfect candidate for `bit_view`.

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -603,6 +603,135 @@ bool DOS_CreateFile(const char* name, FatAttributeFlags attributes,
+{
+	uint8_t data = 0;
+	bit_view<0, 3> access_mode;
+	bit_view<4, 3> sharing_mode;
```
Well, this still leaves you with bit patterns when you interpret these values 😄 

What I meant:

- Put a link to this page above the union: https://stanislavs.org/helppc/int_21-3d.html
- Add bit-level mappings, naming each bit descriptively, just like I started doing in the VGA header files.

That would make the structure self-describing instead of relying on comments and/or magic bit pattern constants (or hex, which can be even worse).

weirddan455 replied:
There's not a meaning for a single bit though.  These are both 3 bit patterns, as you can see from that link you just sent.  There's not like a read bit and a write bit.

johnnovak replied:
Mea culpa. I did not read it in detail and I'm still half asleep 😅 Ok then, just putting the link there would suffice 😄

weirddan455 replied:
Added the link

FeralChild64 reviewed (commented):
It would be nice if the code could log (`LOG_WARNING`) unsupported sharing modes if the application tries to use them.

weirddan455 commented:
Fixed a missing region lock check in `DOS_LockFile()`.  You can't lock a region if some other process has already locked it.

Also, not sure if we are handling child process inheritance of file handles correctly.  Documentation for the full file lock (as set by sharing bits in `DOS_OpenFile()` says that child processes should inherit this sharing if the inherit bit is set.

![file-lock](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/312af91c-0208-4ce2-990e-ac6432f4dc0c)

But documentation for region lock says it should **not** be inherited by child processes.  File handle duplication should be handled correctly.  That function simply inserts a new reference to an existing `DOS_File` in the caller's PSP segment.  Child processes I am unclear on.  Need to test and see if something breaks.

![region-lock](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/ef605597-e101-49b5-badf-afebb4edd1b8)

weirddan455 commented:
I've done some testing and updating the top comment with games tested.  Some of the games on #1489 I could not find proper images for but I tested what I could find.

It looks like we are respecting the inherit bit, at least for full file locking.  Partial file locking I am still unclear on if the locking process creates a child process but I cannot find a fail case in a game so I think we can cross that bridge when we come to it.

This is done and ready for review.

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -180,6 +180,28 @@ static bool file_is_locked(const char *file_name, const uint8_t drive, const uin
+	for (int i = 0; i < DOS_FILES; ++i) {
+		// Ignore locks held by the current file handle,
+		// Need to check other handles pointing to the same file.
+		if (i != file_handle && Files[i] && Files[i]->GetDrive() == Files[file_handle]->GetDrive()
+				&& Files[i]->IsName(Files[file_handle]->GetName())) {
+			for (const auto &lock : Files[i]->region_locks) {
+				if (regions_overlap(pos, len, lock.pos, lock.len)) {
+					return true;
+				}
+			}
+		}
```
Readability improvement suggestion:

```cpp
    if (i != file_handle && Files[i]) {
        const auto drive_match =
            (Files[i]->GetDrive() == Files[file_handle]->GetDrive());

        const auto filename_match =
            (Files[i]->IsName(Files[file_handle]->GetName()));

        if (drive_match && filename_match) {
            for (const auto &lock : Files[i]->region_locks) {
                if (regions_overlap(pos, len, lock.pos, lock.len)) {
                    return true;
                }
            }
        }
    }

```

weirddan455 replied:
Yeah that looks nicer.  Changed.

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -1721,3 +1763,46 @@ void DOS_SetupFiles()
 	        z_drive_index, std::make_unique<Virtual_Drive>());
 }
+
+bool DOS_LockFile(uint16_t entry, uint32_t pos, uint32_t len)
```
Could add const

weirddan455 replied:
const'd

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -1721,3 +1763,46 @@ void DOS_SetupFiles()
+
+bool DOS_UnlockFile(uint16_t entry, uint32_t pos, uint32_t len)
+{
+	const uint32_t handle = RealHandle(entry);
```
auto?

weirddan455 replied:
auto'd

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -50,6 +50,136 @@ DOS_File* Files[DOS_FILES] = {};
+			break;
+		// Assume read/write access for the purpose of file locking checks
+		// This should throw an error when the file actually tries to open
+		default:
```
Comment should go inside the `default` branch & fix indentation.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -50,6 +50,136 @@ DOS_File* Files[DOS_FILES] = {};
+			ret.sharing = FileSharingMode::DenyNone;
+			break;
+		// If we somehow get something invalid, assume compatibiliy mode.
+		default:
```
Comment should go inside the default branch & fix indentation.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -50,6 +50,136 @@ DOS_File* Files[DOS_FILES] = {};
+		case FileSharingMode::DenyNone:
+			return existing_file.sharing != FileSharingMode::Compatibility;
+	}
+	// GCC 12 throws a compile error if we don't have a return statement here
```
Remove comment; we do this everywhere already, we don't need to comment on it (plus I believe PVS-Studio would flag this anyway).

weirddan455 replied:
Removed

johnnovak commented on src/dos/dos_files.cpp:
```diff
@@ -50,6 +50,136 @@ DOS_File* Files[DOS_FILES] = {};
+{
+	// Needs a two-way check
+	// Run this function twice with arguments reversed the second time
+	return single_directional_sharing_check(new_file, existing_file) && single_directional_sharing_check(existing_file, new_file);
```
Formatting (line too long)

weirddan455 replied:
Formatted

johnnovak commented on include/dos_system.h:
```diff
@@ -345,7 +345,7 @@ class DOS_Drive {
 	virtual ~DOS_Drive() = default;
 
-	virtual bool FileOpen(DOS_File** file, char* name, uint32_t flags) = 0;
+	virtual bool FileOpen(DOS_File** file, char* name, uint8_t flags) = 0;
```
const while you're at it

weirddan455 replied:
Some of the `FileOpen` functions modify the arguments.

johnnovak commented on src/dos/drive_overlay.cpp:
```diff
@@ -216,7 +216,7 @@ class OverlayFile final : public localFile {
 
 	bool Write(uint8_t * data,uint16_t * size) override {
-		uint32_t f = flags&0xf;
+		uint8_t f = flags&0xf;
```
minor: reformat

weirddan455 replied:
Formatted

johnnovak commented on src/dos/dos_misc.cpp:
```diff
@@ -63,6 +63,10 @@ static Bitu INT2A_Handler(void) {
 
 static bool DOS_MultiplexFunctions(void) {
 	switch (reg_ax) {
+	case 0x1000:
```
Maybe document what `0x1000` stands for?

weirddan455 replied:
I put a comment directly under that line.  This is an interrupt function and all it does is basically return "true" if `SHARE.EXE` is installed.  I don't know that it even has a documented name.  This is the way all interrupt functions look in our codebase.

johnnovak reviewed (approved):
Still looking good @weirddan455, just a few minor style-related comments.

Grounded0 commented:
Thanks for finally implementing this.

Grounded0 commented:
You can expand your testing software with something like Microsoft Word 6.1 for Windows:

https://winworldpc.com/product/microsoft-word/6x

johnnovak commented:
> You can expand your tested software with something like Microsoft Word 6.1 for Windows:
> 
> https://winworldpc.com/product/microsoft-word/6x

Does it need some locking specific functionality, do you know @Grounded0 ?

Grounded0 commented:
@johnnovak 

Yes.

weirddan455 commented:
@Grounded0 did you get that error on this branch?  If not, I can test it real quick.

Grounded0 commented:
No this is 0.81.1.

weirddan455 commented:
Well, this isn't a good sign.  Word's setup.exe crashes with what looks like a divide by zero.  I think this is unrelated to this PR as it's crashing in VGA code.

```
Loguru caught a signal: SIGFPE
Stack trace:
24      0x56553e48cf81 _start + 33
23      0x7f0524846305 __libc_start_main + 133
22      0x7f052484624a /lib/x86_64-linux-gnu/libc.so.6(+0x2724a) [0x7f052484624a]
21      0x56553e48d072 main + 34
20      0x56553e86077f sdl_main(int, char**) + 7039
19      0x56553e4c1996 Config::StartUp() + 22
18      0x56553eadae4e SHELL_Init() + 3870
17      0x56553ead8d6b DOS_Shell::Run() + 1819
16      0x56553ead7b5c DOS_Shell::ParseLine(char*) + 2124
15      0x56553eae813d DOS_Shell::DoCommand(char*) + 445
14      0x56553eb01433 DOS_Shell::ExecuteProgram(std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >) + 1267
13      0x56553eb025ab build/debug/dosbox(+0xc7f5ab) [0x56553eb025ab]
12      0x56553e5be1c4 CALLBACK_RunRealInt(unsigned char) + 84
11      0x56553e48d913 DOSBOX_RunMachine() + 19
10      0x56553e48d863 build/debug/dosbox(+0x60a863) [0x56553e48d863]
9       0x56553ea5fd6a build/debug/dosbox(+0xbdcd6a) [0x56553ea5fd6a]
8       0x56553ea6a275 INT10_SetVideoMode(unsigned short) + 2437
7       0x56553ea6d8c5 build/debug/dosbox(+0xbea8c5) [0x56553ea6d8c5]
6       0x56553e91d5e1 IO_WriteB(unsigned short, unsigned char) + 401
5       0x56553e48d913 DOSBOX_RunMachine() + 19
4       0x56553e48d7fd build/debug/dosbox(+0x60a7fd) [0x56553e48d7fd]
3       0x56553e983338 PIC_RunQueue() + 232
2       0x56553e9d5f13 VGA_SetupDrawing(unsigned int) + 83
1       0x56553e9d3464 setup_drawing() + 8516
0       0x7f052485b050 /lib/x86_64-linux-gnu/libc.so.6(+0x3c050) [0x7f052485b050]
2024-05-25 21:34:37.404 | Signal: SIGFPE
Floating point exception (core dumped)
```

weirddan455 commented:
Except that's not constantly reproducible.  Now I get this error.  I might have to revert that SFT commit I made in an earlier PR.

![image0008](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/1f20fd61-bf86-428a-be8b-0165e647b998)

Grounded0 commented:
I'll test it myself on a test build this Sunday.

Can you provide me a download link to macOS artefact for this branch?

johnnovak commented:
@weirddan455 @Grounded0 How about we use a "standardised" Win3.x base install for all our testing? Our wiki has a lot of helpful hints on setting up Win3.x, but it's a bit sprawling. I'm thinking of doing a base Win3.x install myself, and create a more streamlined "recommended Win3.x setup" wiki page.

Then probably I'll create a few "snapshots" ZIP files:

- Base install only
- S3 and SB16 drivers only
- Additional updates (WinG, etc.)

How does that sound? Ultimately, we need to support a _single_ S3 driver, and a _single_ SB16 driver and get them work 100% reliably, then all games and apps should just work in theory as they only talk to the hardware through the Windows APIs.

Of course, we should pick non-buggy drivers for this 😄

weirddan455 commented:
Oh, I downloaded Word for DOS.  So the SFT commit I made has a regression where it breaks DOS based programs from working inside Windows 3.1 (displaying the general protection fault error box inside Windows).

The VGA crash is a separate issue (reproducible on `main`) that is not consistently reproducible but happens maybe 1 in 4 times you try to launch a DOS program from inside Windows 3.1.

I'll make/update tickets for those.

Grounded0 commented:
Microsoft Word 6.1 for Windows (3.5-1.44mb):

https://winworldpc.com/download/15c2b403-c2b8-c3a9-c5b8-11c3a5c28f13

johnnovak commented:
> happens maybe 1 in 4 times you try to launch a DOS program from inside Windows 3.1.

In general, launching DOS programs in Win3.x is the last of our concerns. "Nice to the have", I guess, but should be right at the bottom of our priority list. Just use DOS 🤷🏻

weirddan455 commented:
>  In general, launching DOS programs in Win3.x is the last of our concerns. "Nice to the have", I guess, but should be right at the bottom of our priority list. Just use DOS 🤷🏻

True.  Everything continues to work in DOS.  It also breaks Window's MS-DOS prompt, which is the easiest way to reproduce the regression.

In any case, Word for Windows does work on this branch

![image0009](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/1fb72039-4dae-40cf-a104-cd4c44c78e9d)

MasterO2 commented:
I thought Windows support was out of scope for Staging?

Grounded0 commented:
Official position seems to be DOS and Windows 3.x (including Win16 and Win32s).

weirddan455 commented:
>  I thought Windows support was out of scope for Staging? At least that's what Staging's web site lists as a non-goal.

It's Windows 9x specifically that's listed as a non-goal.  We're supporting Windows 3.1 stuff.

Grounded0 commented:
File locking is also required by some DOS application software so this also benefits DOS.

weirddan455 commented:
>  File locking is also required by some DOS application software so this also benefits DOS.

I haven't found any yet but `SHARE.EXE` has been a part of MS-DOS itself since at least version 4.0 so this is definitely in scope for our DOS implementation.

The big distinction between Windows 3.1 and Windows 9x is that Windows 3.1 installs and runs on top of our internal DOS implementation whereas Windows 9x wants to install its own version of DOS on top of bare metal (or bare hardware emulation).

This means things like the `mount` command won't work and maybe even `localDrive` won't work.  I haven't tried it myself but I think you would have to create a FAT image of the Windows 9x install and then use the `boot`` command to boot into it (as far as Staging is concerned `boot` is meant for bootable floppy images of games).

It's not so user friendly to do all that and would require a lot of work to make it so (something like DOSBox-X's GUI).

weirddan455 commented:
@johnnovak A Windows install would be helpful.  I was running into problems with the SB16 drivers myself.

johnnovak commented:
> I thought Windows support was out of scope for Staging? At least that's what Staging's web site lists as a non-goal.

Read the about page carefully, man 😄 Everythig you need to know is there 😄 
https://www.dosbox-staging.org/about/

Grounded0 commented:
@sduensin 

What was the name of the DOS software that required file locking support (SHARE)?

johnnovak commented:
> True. Everything continues to work in DOS. It also breaks Window's MS-DOS prompt, which is the easiest way to reproduce the regression.

Yeah, and the guiding principle to everything we do in DOSBox should be this: does this allow to run the DOS and Win 3.x catalogue in an authentic way? DOS prompt in Win 3.x is an extreme niche; making it work brings no value in light of this goal.

Similarly, it's completely beside the point to get _all_ the possible driver working in Win 3.x for all the hardware we support. We just need a _single_ supported driver combo to run all the games optimally, job done. This allows us to spend our energies in a smart way on what truly matters: get the existing DOS/Win3.x catalogue from 1981-2000 working optimally. Running the same game with 10123 Win3.x driver combos that result in the exact same end-user experience is a non-goal and a waste of time and energy.

Just talking in genericsl I'm not saying we're doing this type of busy work now 😄 

> The big distinction between Windows 3.1 and Windows 9x is that Windows 3.1 installs and runs on top of our internal DOS implementation whereas Windows 9x wants to install its own version of DOS on top of bare metal (or bare hardware emulation).

Win 95 and 98 is a huge rabbit hole, and we have enough to in pure DOS/Win3.x land. Hence, we _really_ don't wanna go there, IMO. Keep the scope realistic and all.

Win 9.x kinda sorted "worked" (works) in all DOSBox version by fluke and random coincidence, but to make it work "properly", there's lots of additional work to do. Imagine dealing with all Win9.x game compatibility tickets then... Just no, we don't even wanna go there IMO. We are quite under-resourced just for the DOS / Win3.x stuff, we don't wanna take on something absolutely huge as Win 9.x (to do it properly, not just half-assedly).

Grounded0 commented:
SHARE.EXE has been present in MS-DOS at least since MS-DOS 3.1 (search for SHARE.EXE):

http://www.ctyme.com/intr/int-21.htm

weirddan455 commented:
>  We are quite under-resourced just for the DOS / Win3.x stuff, we don't wanna take on something absolutely huge as Win 9.x (to do it properly, not just half-assedly).

I agree 100%.  I'm not even sure DOSBox is the right approach at all for Win9x.  Maybe DOSBox-X does a decent job, haven't tried it but all the DOS stuff we implement becomes useless and then you're also running an entirely different class of hardware.  A lot of games need Pentium 2/3 class CPUs and much stronger GPUs than Voodoo.

johnnovak commented:
> I agree 100%. I'm not even sure DOSBox is the right approach at all for Win9x. Maybe DOSBox-X does a decent job, haven't tried it but all the DOS stuff we implement becomes useless and then you're also running an entirely different class of hardware. A lot of games need Pentium 2/3 class CPUs and much stronger GPUs than Voodoo.

Absolutely, even the emulated hardware requirements are significantly different, that's an excellent point. There is overlap, but I'd say Win 9x emulation should be a distinct and separately project from DOSBox; it's really its own thing. Huge respect to DOSBox-X for trying to pull it off, and all the best wishes in the world & good luck, but this is not for us.

> A Windows install would be helpful. I was running into problems with the SB16 drivers myself.

I'll do that over the coming days, ideally sometime next week. The planned outcome is a new succinct wiki page with installation steps, and a couple of ZIP archives on archive.org if someone just wants to quickly grab it and not go through the steps.

weirddan455 commented:
>  I'll do that over the coming days, ideally sometime next week.

Sounds good.  Anyway, I've addressed your comments so I'll merge soon if no objections.

johnnovak commented:
> > I'll do that over the coming days, ideally sometime next week.
> 
> Sounds good. Anyway, I've addressed your comments so I'll merge soon if no objections.

Thanks! Merge away, man 😎

interloper98 commented:
This use mentions that their accounting software's multi-user SQL back-end doesn't work under original dosbox, which might point to the lack of share support:

https://www.vogons.org/viewtopic.php?f=7&t=37333&p=328571&hilit=needs%20need%20share%20exe#p328571

My hunch is that it's Peachtree Accounting, because they were massive in the early 2000's -- even their 2004 release of Peachtree Complete Accounting 11 was solely DOS based!

I could only find v 8.0 on archive: https://archive.org/details/peachtree-complete-accounting-8.0-2001, but I haven't tried it w/ the latest CI builds yet.

interloper98 commented:
![8rn448](https://github.com/dosbox-staging/dosbox-staging/assets/165201780/a89cd344-89ab-449b-9924-06af7838a13f)

weirddan455 commented:
The DOSBox Staging team will not be held liable for any and all accounting errors.

Actually I think we've had to make a statement like that before.  I vaguely remember it came out that someone was using DOSBox for some engineering software and our floating-point calculations are not always 100% accurate on all platforms compared to old x86 hardware that uses 80-bit precision.

So yeah, we won't be held liable for any collapsed bridges either.

johnnovak commented:
> The DOSBox Staging team will not be held liable for any and all accounting errors.
> 
> Actually I think we've had to make a statement like that before. I vaguely remember it came out that someone was using DOSBox for some engineering software and our floating-point calculations are not always 100% accurate on all platforms compared to old x86 hardware that uses 80-bit precision.
> 
> So yeah, we won't be held liable for any collapsed bridges either.

[I've already taken care of this.](https://www.dosbox-staging.org/about/#disclaimer)

johnnovak commented:
@Python-Exoproject Just pinging you to show off the rapid progress we're making on the Win 3.x emulation front 😎

interloper98 commented:
Peachtree Complete Accounting 9.0 runs perfectly.

The startup batch file (`PEACH.BAT`) first runs a tool call `SHARETST.EXE` and from its return code decides if it should run launch the local/stand-alone `PT.EXE` or the network executable `NETPEACH.EXE`.

I suspect the `SHARETST.EXE` can detect if the files are sitting on a network share or are local. (The old way of running this software in a multi-tenant environment is that you have a Novell or NetBUI server hosting the sole copy of the software, and lots of accountants client machines mounting this shared directory. Lots of users can then be working simultaneously, and Peachtree takes care of the serializing/locking/unlocking access behind the scenes).
 
The main Staging CI build and DOSBox-X can run both executables:


https://github.com/dosbox-staging/dosbox-staging/assets/165201780/b21a1b98-1452-489b-9665-c9388a4652b3


https://github.com/dosbox-staging/dosbox-staging/assets/165201780/a72a9b27-9288-4b92-996c-b3d60064dcd5

But older 81.1 can't run them. Not even `FAKESHAR` is enough to fool it:


https://github.com/dosbox-staging/dosbox-staging/assets/165201780/ac503403-186d-417d-8880-e3da72b54b72

Unfortunately I'm not too sure how to get Microsoft's NetBUI client working (and same w/ Novel's 2.7 client + NE2000 drivers). So I was able to test a real DOS-based network share. 

@Torinde - do you know how to share and mount a DOS directory across two DOSBox instances (either Ethernet or IPX, using either MS or Novel?)
